Enhance: allow the use of system provided QtKeyChain packages (both QMAKE & CMAKE)

SlySven opened the pull request:
If environmental variable `WITH_SYSTEM_QTKEYCHAIN` is defined to (Case insensitive) `YES` or `QT5` then a system package which places the relevant files in `qt5keychain` folders will be used; if it is instead defined to `QT` then files in `qtkeychain` folders will alternatively be used. If said files are not present or the wrong ones requested then builds will fail both during compilation and linking. If the variable is set to any other value or not defined at all then Mudlet will do as it did before and download the library from a git repository - there is a separate PR in progress at the moment to change where from and that might need revision to combine cleanly with this PR.

~~Note that THIS commit only addresses the QMake build system. Further work is needed for the CMake one and that t~~ *T*his WILL clash/not combine with PR https://github.com/Mudlet/Mudlet/pull/3603 .

Also note that the semantics are not quite the same as other build time environment options - because there seems to be more than one possible system/directory naming system for this option. However the build behaviour, if this option is NOT invoked, should be the same as before.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: allow the use of system provided QtKeyChain packages (QMAKE ONLY)

If environmental variable `WITH_SYSTEM_QTKEYCHAIN` is defined to (Case
insensitive) `YES` or `QT5` then a system package which places the relevant
files in `qt5keychain` folders will be used; if it is instead defined to
`QT` then files in `qtkeychain` folders will alternatively be used. If
said files are not present or the wrong ones requested then builds will
fail both during compilation and linking. If the variable is set to any
other value or not defined at all then Mudlet will do as it did before and
download the library from a git repository - there is a separate PR in
progress at the moment to change where from and that might need revision
to combine cleanly with this PR.

Note that THIS commit only addresses the QMake build system. Further work
is needed for the CMake one and that this WILL clash/not combine with
PR https://github.com/Mudlet/Mudlet/pull/3603 .

Also note that the semantics are not quite the same as other build time
environment options - because there seems to be more than one possible
system/directory naming system for this option. However the build
behaviour, if this option is NOT invoked, should be the same as before.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: allow use of system provided QtKeyChain packages (CMAKE addendum)

This *seems* to work but the behaviour if `WITH_SYSTEM_QTKEYCHAIN` is not
set to anything needs confirmation on the CI platforms...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: try and repair broken CI CMake Linux & MacOs builds

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: correct a CMake typo in last commit area

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: a typo that causes a link failure

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: improve use of environmental variable WITH_SYSTEM_QTKEYCHAIN

Due to some errors in the use or not of `$` in some places things were not
working as expected. As some extra debugging messages revealed.

Also cleared up a misunderstanding about the behaviour of CMake
add_subdirectory(...) function.

After reviewing the prior attempt to do some of what this PR does I
determined that `find_package(Qt5Keychain)` should also work for the
system provided QtKeychain library - and return a Qt5Keychain_FOUND boolean
when it does. It also seems that it should be possible to determined
whether the library is located in a `qt5keychain` or `qtkeychain`
sub-directory and to report if it does not match our expectations and thus
cause a mismatch in the relevant `#include`s.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Faenriis commented:
It's still hardcoded in discord.h
#include "../3rdparty/discord/rpc/include/discord_register.h"
#include "../3rdparty/discord/rpc/include/discord_rpc.h"
We need to figureout how to solve those hardcoded paths first, otherwise qmake/cmake config is futile.

Xabre666 commented:
Totally mixed what I was workin on, idiot..
dlgConnectionProfiles.cpp:#include <../3rdparty/qtkeychain/keychain.h>
dlgConnectionProfiles.h:#include <../3rdparty/qtkeychain/keychain.h>
Stil hardcoded in those files, we need to fiure out how to solve hardcoded include paths in the source first.
Good effort, though.

Xabre666 commented:
Seems to me that this could be solved in slightly different manner. Mudlet could still use code from submodule, but install paths for submodule could be specifically set to a location that is dedicated to Mudlet files. That way it simply would not matter how many copies of some submdoule like qtkeychain you have on your system, Mudlet will alway use its version and paths.
Reading this atm: https://cmake.org/cmake/help/v3.2/module/ExternalProject.html

Xabre666 commented:
Yes, seems that set(ExternalProjectCMakeArgs) macro can be used to set specific args to define install paths for submoule code.
Talking only about Cmake build, since QMake is not an issue here! When building with qmake, everything ends up bundled up into a single mudlet executable. CMake build on the other hand works with multiple files instead, and that's where file conflicts are created.

SlySven commented:
:confounded: Humm, the `find_package(Qt5Keychain REQUIRED)` that failed in the CI builds was intended to find the Mudlet self-built ones - and it worked for me.

@Faenriis In **your** case what value did you have the needed override `WITH_SYSTEM_QTKEYCHAIN` environmental value set to? If it was set to a case insensitive `YES` or `QT5` those includes in `mudlet.h` and `dlgConnectionProfiles.h` will be switched to the needed `<qt5keychain/keychain.h>`, or if set to `QT` they will be switched to `<qtkeychain/keychain.h>`. **In *ALL* cases the bogus `#include` in `dlgConnectionProfiles.cpp` was removed in commit https://github.com/Mudlet/Mudlet/commit/deecc82f1f4d37decd119b722afab5f2b1882d5a** so I do not understand WHAT code you are using in https://github.com/Mudlet/Mudlet/pull/3619#issuecomment-614456444 ? :confused:

SlySven commented:
Humm, it seems that `find_package(...)` is not a good idea when Mudlet wants to use it's own copy, certainly, *before* it has been built in a CI environment...  have to make use of that only when we need it to find a *system* package!

SlySven committed:
Revise: only use find_package(Qt5Keychain) for system library case

It seems we should not use this CMake command when we are building the
library ourselves...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Testing: add extra debug output to travis.osx.install.sh [skip appveyor]

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: some more debugging tweaks [skip appveyor]

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: add zziplib to qmake pkg-config packages [skip appveyor]

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Test: add pkg-config output for zziplib to MacOs CI install [slip-appveyor]

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: cure typo in previous commit

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Debug: actually list MacOS contents of dir for zziplib lib [skip appveyor]

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: bodge needed links for MacOs libzzip library [skip appveyor]

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: fix typo and make symbolic links with sudo [skip appveyor]

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 committed:
Merge branch 'development' into Enhance_useSystemQtKeychainIfRequested

SlySven committed:
Revise: undo cosmetic changes to CMakeLists.txt & cleanup fixing CI builds

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

keneanung reviewed (changes requested):
Not happy with the cmake changes at all. It adds another new environment variable semantics and puts conditionals all over the place. We should be able to abstract all of this away into an own module that defines a common target for all cases. I'm happy to attempt that in a PR against this branch that covers the module.

For testing, do you have an example distribution where the make module is called `qtkeychain`? That would make testing much easier.

Second: Is the manual choice of `qtkeychain` vs `qt5keychain` important or is it something that can be delegated to a machine? If it could be delegated, maybe the environment variable concept could be made the same as with the other options?

Xabre666 commented:
Ok, finally able to test this, sorry for the delay.

Build using qmake/make - success, resulting in working mudlet that uses system qtkeychain with no conflicting files.

Build using cmake/make- fails.
[cmake_make-faillog.txt](https://github.com/Mudlet/Mudlet/files/4512246/cmake_make-faillog.txt)

Thus declaring this an overall failure at this state, as the sole purpose of opting to use system provied qtkeychain is to avoid file conflicts when building via cmake/make. Due to the nature of qmake build (resuting Mudlet excutable contains all relevant libraries within itself, thus no file conflicts), succesful qmake build means nothing - it was working without resulting file conflicts even before these changes.

On that note, I have to agree with @keneanung , I feel that this is being done in slightly overcomplicted way that adds more environmental variables and relies on automatic detection and just adds more code to maintain a hope it'll work as system changes are introduced.

What puzzles me more is that different aproach has already being tried and utilized for another  git submodule : lua code formater. If you check that code, lcf is is not bulit using default build flags defined in submodule, but more specific, more suitable install path is being set for the relvant files, path that is reserved for Mudlet's files and thus creates no conflicts (in given case, system installation of lcf conflicts in no way, system and Mudet's version coexist in harmony).
So,the question is: why not use the same approach for all other submodules?

SlySven commented:
The Lua Code Formatter is not needed during the build process - it is something that gets loaded in at run time and if the needed resources are not present then, Mudlet moans about it but otherwise carries on without it.

A C/C++ library that is linked into the source code is a significantly more unwieldy beasty, it needs header files to be in an expected location during compilation - that location for headers inside `#include <`...`>` is relative to the "system" directories which for CMake are stored in the `CMAKE_SYSTEM_INCLUDE_PATH` variable but supplemented by what gets put into `CMAKE_INCLUDE_PATH` by various `find_xxx` operations and other things during the `cmake` process. Therefore to change that to switch between a system header file and one internal to our project in an unusual place we have to change the `#include`  lines - and it is not easy to do that as the C/C++ preprocessor does not allow that line to contain preprocess *variables* as I understand it. The simplest and most reliable method is to select from one of a limited set with a `#if xxx` ... `#include <ABC>` ... `#elif yyy` ... `#include <DEF>` ... `#else` ... `#include <GHI>` ... `#endif` form.

Including the right library at the final linking stage is a bit less convoluted but it can be a bit more cryptic as instead of the compiler moaning about not finding a header file where it is referred to in the source code (as you get in the previous paragraph) you just get undefined references to functions that you may not even recognise...

As for the need to have two different internal *switches* that is because the library was known as `qtkeychain` but at least one big distribution Debian (and thus by implication Ubuntu) made a switch when they had packages for both Qt4 and Qt5 side-by-side 0- hence things ending up in with a name / path of `qt5xxx` instead of `qt` - whilst nowadays they no longer maintain Qt4 stuff they haven't changed the name back the one without the `5` however ***I cannot determine that other distributions have done the same thing*** - so it made sense to try and accommodate either that might arise.

I would be willing to work with you @Faenriis to see what is going wrong and to try and fix it with you.

SlySven committed:
Revise: simplify now we know that we only have to deal with Qt5 version

The builds of QtKeyChain without a 5 in the include paths is the older one
that is built against Qt4 - since we do not work with anything less than
Qt 5.11 we can ignore that variant which means we can revert to the
methodology we use for other options - which simplifies the CMake case a
bit!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: invert sense of & rename peration of environmental variable

This is to confirm with the sense of operation with our other build
configuration environmental variables and this one has been renamed to
`WITH_OWN_QTKEYCHAIN` which MUST be a case insensitive `NO` to prevent the
use of the embedded copy of the library.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

keneanung commented on src/CMakeLists.txt:
```diff
@@ -323,6 +321,21 @@ if(Qt5Gamepad_FOUND)
+if(USE_OWN_QT5KEYCHAIN)
+  # An extra definition is needed if we are using our own QtKeyChain
+  target_compile_definitions(mudlet PRIVATE INCLUDE_OWN_QT5_KEYCHAIN QTKEYCHAIN_NO_EXPORT)
+  message(STATUS "Using own QtKeyChain (Qt5) library.")
```
This message is already output by the `include_optional_module` helper, so I'd skip it here.

SlySven replied:
Fair comment but it does seem like we get two sets of status messages put out and I am not sure that the first batch gets repeated when the build process is repeated, unless the whole CMake configuration of the project gets blown away first. I have another look and see...

SlySven replied:
.. okay it has been removed.

keneanung commented on src/CMakeLists.txt:
```diff
@@ -323,6 +321,21 @@ if(Qt5Gamepad_FOUND)
 
+if(USE_OWN_QT5KEYCHAIN)
+  # An extra definition is needed if we are using our own QtKeyChain
+  target_compile_definitions(mudlet PRIVATE INCLUDE_OWN_QT5_KEYCHAIN QTKEYCHAIN_NO_EXPORT)
```
I'd move this to the other `target_compile_definitions` above.

SlySven replied:
Those above ones are unconditional whereas ~this one is~ both of them are dependent on being used for only for the built in QtKeychain - indeed the first one is what controls this feature in the compilable code.

keneanung commented on src/CMakeLists.txt:
```diff
@@ -323,6 +321,21 @@ if(Qt5Gamepad_FOUND)
+  message(STATUS "Using own QtKeyChain (Qt5) library.")
+else()
+  # Only use this when we are using a system package and NOT our own local build
+  find_package(Qt5Keychain REQUIRED)
```
Can we have this near the other `find_package` calls?

keneanung commented on src/CMakeLists.txt:
```diff
@@ -323,6 +321,21 @@ if(Qt5Gamepad_FOUND)
+  message(STATUS "Using system QtKeyChain (Qt5) library (found version \"${Qt5Keychain_VERSION}\").")
+endif()
+
+target_link_libraries(mudlet qt5keychain)
```
This can be added back to the list of linked libraries if the calls above are moved as requested.

keneanung reviewed (commented):
This makes me much happier :slightly_smiling_face: A few structural comments though.

Xabre666 commented:
hmm
_LUA_INCLUDE_DIR=/usr/include/lua5.1 WITH_FONTS=NO WITH_UPDATER=NO WITH_OWN_QTKEYCHAIN=NO cmake -DCMAKE_INSTALL_PREFIX=/opt/mudlet ..

-- Excluding optional updater module explicitly
-- Excluding optional fonts module explicitly
-- Including optional 3D mapper module
-- Including optional build-type splash screen module
-- Excluding optional own QKeychain (Qt5) library module explicitly
_

But when CMake actually finishes
_-- Using TextToSpeech module
-- Using Gamepad module
-- Using own QtKeyChain (Qt5) library.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/slobo/Mudlet/build_

Starts by stating that qtkeychain is excluded explicity but ends up using own qtckechain library.

keneanung commented:
Did you delete your CMakeCache.txt?

Xabre666 commented:
Yeah, that was it, stupid me
_Using system QtKeyChain (Qt5) library (found version "0.10.0")_
Sorry for alarm :(

SlySven committed:
Revise: peer-review suggestions

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: correct a stupid typo

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

keneanung commented on src/CMakeLists.txt:
```diff
@@ -311,7 +326,8 @@ target_link_libraries(
   ZIP::ZIP
-  ZLIB::ZLIB)
+  ZLIB::ZLIB
+  qtkeychain)
```
This failed the build. Suggested change to fix it:
```suggestion
  qt5keychain)
```

SlySven replied:
:facepalm:

keneanung reviewed (commented):
Well, compilation broke due to another typo, but otherwise this looks much better. Thanks :slightly_smiling_face:

SlySven committed:
Revise: another stupid typo

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: add a library include needed just for FreeBSD

Whilst this is not essential for the main line code it helps to fix one
thing on FreeBSD and is the CMake counterpart to an existing PR
https://github.com/Mudlet/Mudlet/pull/3656 .

This being done here purely to put another commit onto this PR to try and
get the previously failed GitHub Action that is a Ubuntu CI build to
run successfully this time. The failure mechanism then was related to the
provisioning of the CI platform and NOT the code being tested.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

keneanung commented on src/CMakeLists.txt:
```diff
@@ -329,6 +329,10 @@ target_link_libraries(
   qt5keychain)
 
+if(CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+  target_link_libraries(mudlet sysinfo)
```
Is this library addition part of the qtkeychain change?

SlySven replied:
No, it was something I needed to go in anyhow and I wanted something to force a rebuild on the GH Actions Ubuntu platform as I couldn't see any other way to force a rerun/rebuild.  I am not sure but it is possible that Vadim might have plans to put a button on the Actions user interface that can do it but that is not present here, now {see https://www.actionspanel.app/ }.

SlySven replied:
It is actually the CMake equivalent to #3656 .

SlySven replied:
Also I didn't spot the reason that the PR wasn't merge-able was peer-review and not broken CI.

vadi2 replied:
I'm not sure we can have a button to force a rebuild, besides the one that Github shows if a build failed: https://help.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run#viewing-your-workflow-history see "Re-run all checks"

SlySven replied:
Just found that - it wasn't quite where I was expecting - or at least I think it wasn't before.

SlySven replied:
It doesn't work!
![Screenshot_20200424_prior_to_clicking_restart](https://user-images.githubusercontent.com/6163092/80173536-1174de00-85e8-11ea-98a2-2c59a714378e.png)

You can click it and the dark part of the screen (containing the log) goes blank apart from a flashing octo-cat head and you get a green banner at the top saying that the job restarted
![Screenshot_20200424_after_clicking_restart](https://user-images.githubusercontent.com/6163092/80173535-1174de00-85e8-11ea-97ae-7df93673877f.png)

But the button that you click still says "click to restart job".
![Screenshot_20200424_30sec_or_more_after_clicking_restart](https://user-images.githubusercontent.com/6163092/80173534-10dc4780-85e8-11ea-91e8-20462141f0ae.png)

After 30 seconds of no change I clicked it again and this time got a red banner at the top saying there was a problem requesting jobs from GH Accounts:
![Screenshot_20200424_after_clicking_restart_AGAIN](https://user-images.githubusercontent.com/6163092/80173532-1043b100-85e8-11ea-88f5-a839ffbbeae9.png)

Shortly after that I finally did a refresh of the browser tab - and then found that there looked to have been no change - it was still reporting the details from many hours ago:
![Screenshot_20200424_after_refreshing_screen](https://user-images.githubusercontent.com/6163092/80173530-0fab1a80-85e8-11ea-8fff-a6c028d0eb84.png)

SlySven replied:
Interesting to note that the coloured logo initially showed 2/3 green and 1/3 red - which suggests that two runs of the MacOs job had completed by then and one Utubuntu failure but by the end it had changed to 1/4 red (1x failed Ubuntu job), 1/4 orange (1x still in progress Ubuntu) and 1/2 green (2x good MacOS) and the button is now RED saying cancel workflow - and it only did that AFTER I refreshed the screen:
![image](https://user-images.githubusercontent.com/6163092/80174283-18045500-85ea-11ea-8a00-dc2c973e7f8d.png)

keneanung replied:
Please remove this change then, so we keep PRs single topic :slight_smile:

SlySven replied:
Done, BTW.

SlySven committed:
Revert "Revise: add a library include needed just for FreeBSD"

This reverts commit f40b9d44a9f8ee3a1dc8e37c817ad146a43ef3b5 so the change
can be refactored out to a separate PR.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_useSystemQtKeychainIfRequested

Needed to bring in some changes to the CI infrastructure to enable
this PR to pass them all.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

keneanung reviewed (approved):
Looks good to me now. Let's hope it fulfils the package maintainer's needs.
